### PR TITLE
[FFmpeg] Add .dsf to extension list (DSD support)

### LINKED ISSUE #796: DSD support
Is there any way to get cmus to play DSD files?

### PR DESCRIPTION
**Pull request by gavtroy**
\+ resample to a lower sample rate if needed, as some variants of DSD exceed the 524kHz we support internally -- and the 384kHz limit of at least some outputs (pulse).

Fixes #796.

### CODE REVIEW

File: ip/ffmpeg.c [hunk 1]
```diff
@@ -248,15 +250,15 @@ static int ffmpeg_open(struct input_plugin_data *ip_data)
...
 	ip_data->private = priv;
 #if LIBAVCODEC_VERSION_MAJOR >= 58
-	ip_data->sf = sf_rate(cc->sample_rate) | sf_channels(cc->ch_layout.nb_channels);
+	ip_data->sf = sf_rate(out_sample_rate) | sf_channels(cc->ch_layout.nb_channels);
 #else
-	ip_data->sf = sf_rate(cc->sample_rate) | sf_channels(cc->channels);
+	ip_data->sf = sf_rate(out_sample_rate) | sf_channels(cc->channels);
 #endif
 	switch (cc->sample_fmt) {
 	case AV_SAMPLE_FMT_U8:
```
Comment (gavtroy) [inline ip/ffmpeg.c:264]: By the way it looks like `default`/`AV_SAMPLE_FMT_S16` is the only one that actually plays back normally. Removing the other cases works but I'd like to understand it more.

(This is nothing to do with DSD or this change though. DSD is `AV_SAMPLE_FMT_FLTP` so it takes the `default` path without issue.)

**Comment by pgaskin** [comment]
@gavtroy, are you still working on this?

File: ip/ffmpeg.c (hunk 1 above)
Comment (gavtroy) [reply ip/ffmpeg.c:264]: The issue with the `AV_SAMPLE_FMT_U8` and `AV_SAMPLE_FMT_S32` cases is the hardcoded sizeof(int16_t) in the buffer_used_len calculation.

**Comment by gavtroy** [comment]
I looked into why the limit was 384000 and the resampling was actually working fine, the limit was from the output side (pulse).

Do note that although this seems to work fine I'm not absolutely confident if it's "right". Forcing a higher sample rate to a lower sample rate like this behaves completely normally, but doing the opposite is clearly broken - the audio cuts off early and seeking is a bit weird. I think it probably just breaks some assumption we make about the output not growing and that the high->low case is fine, but ideally I'd understand it properly and right now I just don't have the energy to do that. In any case we only downsample when it's >384000 so this isn't relevant for pretty much anything except DSD128.

**Review by pgaskin — APPROVED** [review]
Code makes sense, shouldn't cause regressions, and seems to work fine with DSD on AAudio and PulseAudio.

> I think it probably just breaks some assumption we make about the output not growing and that the high->low case is fine, but ideally I'd understand it properly and right now I just don't have the energy to do that

It's not apparent to me either, but this works, and why higher sample rates don't is something we can figure out later, probably when I implement #1328.

> In any case we only downsample when it's >384000 so this isn't relevant for pretty much anything except DSD128.

Agreed. This change doesn't affect lower sample rates, and ones higher than this are rarely seen in mainstream formats.

**Comment by pgaskin** [comment]
@flyingmutant, this should be rebase-merged since the sample size fix commit is a separate (but related) bugfix.

**Pull request merged**